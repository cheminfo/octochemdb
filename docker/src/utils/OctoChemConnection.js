import delay from 'delay';
import dotenv from 'dotenv';
import md5 from 'md5';
import { MongoClient } from 'mongodb';

import debugLibrary from './Debug.js';

dotenv.config();
const debug = debugLibrary('OctoChemConnection');

export function OctoChemConnection() {
  this.client = new MongoClient(process.env.MONGODB_URL, {
    keepAlive: true,
    connectTimeoutMS: 6 * 60 * 60 * 1000,
    socketTimeoutMS: 6 * 60 * 60 * 1000,
  });
}

OctoChemConnection.prototype.close = async function close() {
  if (this.connection) return this.connection.close();
  return undefined;
};

OctoChemConnection.prototype.getCollectionNames =
  async function getCollectionNames() {
    return (await (await this.getDatabase()).listCollections().toArray()).map(
      (entry) => entry.name,
    );
  };

OctoChemConnection.prototype.getCollection = async function getCollection(
  collectionName,
) {
  return (await this.getDatabase()).collection(collectionName);
};

OctoChemConnection.prototype.getImportationLog =
  async function getImportationLog(options) {
    const { collectionName, sources, startSequenceID } = options;
    const logsCollection = await this.getImportationLogsCollection();
    const sourcesHash = md5(JSON.stringify(sources));
    const _id = sourcesHash;
    let logs = await logsCollection.find({ _id }).next();
    if (logs === null) {
      logs = {
        _id,
        collectionName,
        sources,
        sourcesHash,
        dateStart: Date.now(),
        dateEnd: Date.now(),
        startSequenceID,
        endSequenceID: 0,
        status: 'updating',
      };
      await logsCollection.insertOne(logs);
    }
    return logs;
  };
OctoChemConnection.prototype.updateImportationLog =
  async function updateImportationLog(logs) {
    const collection = await this.getImportationLogsCollection();
    logs.dateEnd = Date.now();
    await collection.replaceOne({ _id: logs._id }, logs);
  };

OctoChemConnection.prototype.getImportationLogsCollection =
  async function getCollection() {
    return (await this.getDatabase()).collection('importationLogs');
  };

OctoChemConnection.prototype.getAdminCollection =
  async function getAdminCollection() {
    return this.getCollection('admin');
  };

OctoChemConnection.prototype.setProgress = async function setProgress(
  progress,
) {
  const collection = await this.getAdminCollection();
  await collection.replaceOne({ _id: progress._id }, progress);
};

OctoChemConnection.prototype.getProgress = async function getProgress(
  collectionName,
) {
  const adminCollection = await this.getAdminCollection();
  const _id = `${collectionName}_progress`;
  let progress = await adminCollection.find({ _id }).next();
  if (progress === null) {
    progress = {
      _id,
      state: 'updating',
      seq: 0,
      dateStart: Date.now(),
      dateEnd: 0,
    };
    await adminCollection.insertOne(progress);
  }
  return progress;
};

OctoChemConnection.prototype.getDatabase = async function getDatabase() {
  while (true) {
    try {
      await this.init();
      break;
    } catch (e) {
      debug('Connection to mongo failed, waiting 5s');
      debug(e.stack);
      await delay(5000);
    }
  }
  return this.connection.db(process.env.MONGO_DB_NAME);
};

OctoChemConnection.prototype.init = async function init() {
  if (this.connection) return;

  debug(`Trying to connect to: ${process.env.MONGODB_URL}`);

  this.connection = await this.client.connect();
  debug('Got DB connection');
};
