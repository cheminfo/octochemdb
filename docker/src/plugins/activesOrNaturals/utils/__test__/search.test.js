import { describe, it, expect } from 'vitest';

import { OctoChemConnection } from '../../../../utils/OctoChemConnection.js';
import entriesSearch from '../../routes/v1/search.js';

describe('search activesOrNaturals', async () => {
  const connection = new OctoChemConnection();

  while (true) {
    const activeOrNaturalsCollection =
      await connection.getCollection('activesOrNaturals');
    if ((await activeOrNaturalsCollection.countDocuments()) === 63) {
      break;
    }
  }
  it('search: em using limit', async () => {
    let request = {
      query: {
        em: '344.1623, 376.1647',
        precision: '100',
      },
    };
    const results = await entriesSearch.handler(request);

    request.query.limit = 2;
    const resultsTwo = await entriesSearch.handler(request);
    expect(results.data.length).toBeGreaterThan(resultsTwo.data.length);
    const resultToMatch = results.data.filter(
      (entry) =>
        entry._id ===
        'fi{AP@@QrtZTjjsJk]jxHyHRuUUUUUUUU@AjxVIXUcVCXmavOXccNJx{b^ExQFMkX~FOaaul_CGpp',
    )[0];
    expect(resultToMatch).toMatchInlineSnapshot(`
      {
        "_id": "fi{AP@@QrtZTjjsJk]jxHyHRuUUUUUUUU@AjxVIXUcVCXmavOXccNJx{b^ExQFMkX~FOaaul_CGpp",
        "data": {
          "em": 344.16237387137,
          "mf": "C20H24O5",
        },
      }
    `);
  });

  it('search: min max and flags', async () => {
    const request = {
      query: {
        minNbMassSpectra: 0,
        maxNbMassSpectra: 2,
        minNbActivities: 1,
        maxNbActivities: 2,
        minNbTaxonomies: 1,
        maxNbTaxonomies: 2,
        minNbPatents: 0,
        maxNbPatents: 2,
        minNbPubmeds: 0,
        maxNbPubmeds: 2,
        isNaturalProduct: true,
        fields: 'data.mf,data.em,data.taxonomies,data.activities',
      },
    };
    const results = await entriesSearch.handler(request);
    const resultsTwo = await entriesSearch.handler({
      query: { isBioactive: false },
    });
    expect(resultsTwo.data.length).toBeLessThan(results.data.length);
    const resultToMatch = results.data.filter(
      (entry) =>
        entry._id ===
        'ehZPL@@@KglbbdRebTLRdJttTTRxDLlbRZzv~jjjjjjjjjjjjP@MTClWXRCKjfOacxX',
    )[0];
    expect(resultToMatch).toMatchSnapshot();
  });
  it('search: keywords ', async () => {
    const request = {
      query: {
        kwTaxonomies: 'viridiplantae',
        kwBioassays: 'inhibitors',
        kwActiveAgainst: 'borrarchaeaceae',
        fields:
          'data.kwTaxonomies,data.kwBioassays,data.kwActiveAgainst,data.kwMeshTerms',
      },
    };
    const results = await entriesSearch.handler(request);
    expect(results.data[0].data).toMatchInlineSnapshot(`
      {
        "kwActiveAgainst": [
          "archaea",
          "borrarchaeaceae",
          "borrarchaeales",
          "borrarchaeia",
          "borrarchaeota",
          "borrarchaeum",
          "candidatus",
        ],
        "kwBioassays": [
          "eukaryotic",
          "inhibitors",
          "initiation",
          "molecule",
          "small",
          "translation",
          "uhts",
        ],
        "kwTaxonomies": [
          "viridiplantae",
          "streptophyta",
          "magnoliopsida",
          "cucurbitaceae",
          "momordica",
          "charantia",
        ],
      }
    `);
  });
  it('search: mf ', async () => {
    const request = {
      query: {
        mf: 'C20H28O4',
        fields: 'data.mf,data.em',
      },
    };
    const results = await entriesSearch.handler(request);
    expect(results.data[0]).toMatchInlineSnapshot(`
      {
        "_id": "fak@P@@RuGIEDdeELieEKCbgEAsSMUUUUUUP@ZdCbNBx[c}HYFecxX~F@",
        "data": {
          "em": 332.19875938072,
          "mf": "C20H28O4",
        },
      }
    `);
  });
  it('search: noStereoTautomerID ', async () => {
    const request = {
      query: {
        noStereoTautomerID: 'f`~@P@@HiIImm[Ujjjjj`@upCDLHq~dLATq|L_C@',
        fields: 'data.mf,data.em',
      },
    };
    const results = await entriesSearch.handler(request);
    expect(results.data[0]).toMatchInlineSnapshot(`
      {
        "_id": "f\`~@P@@HiIImm[Ujjjjj\`@upCDLHq~dLATq|L_C@",
        "data": {
          "em": 232.16745925179998,
          "mf": "C12H24O4",
        },
      }
    `);
  });
});
