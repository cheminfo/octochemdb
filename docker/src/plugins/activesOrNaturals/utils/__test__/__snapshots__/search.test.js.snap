// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`search activesOrNaturals > search: min max and flags 1`] = `
{
  "_id": "ehZPL@@@KglbbdRebTLRdJttTTRxDLlbRZzv~jjjjjjjjjjjjP@MTClWXRCKjfOacxX",
  "data": {
    "activities": [
      {
        "$id": "NPC100333",
        "$ref": "npasses",
      },
    ],
    "em": 456.36034540574997,
    "mf": "C30H48O3",
    "taxonomies": [
      {
        "dbRef": {
          "$id": "NPC100333",
          "$ref": "npasses",
        },
        "family": "Fagaceae",
        "genus": "Quercus",
        "kingdom": "Viridiplantae",
        "species": "Quercus acuta",
        "superkingdom": "Eukaryota",
      },
    ],
  },
}
`;
